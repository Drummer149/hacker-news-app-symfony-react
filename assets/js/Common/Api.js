import Firebase from './Firebase';

/**
 * Base API Service to connect to data 
 */
const Api = class ApiService{

  constructor(store = Firebase, version = "v0"){
    this.API_VERSION = version;
    this.DB = Firebase.database();
  };

  // Method to return a promise of an array of story ids from a given endpoint
  getStoryIds(endPoint){
    return new Promise(resolve => {
      this.DB.ref(`${this.API_VERSION}/${endPoint}`).once('value', (storyIds) => {
        resolve(storyIds.val());
      });
    })
  };

  // Method to return a promise of an indivdual item object
  getItem(id){
    return new Promise(resolve => {
      this.DB.ref(`${this.API_VERSION}/item/${id}`).once('value', (item) => {
        resolve(item.val());
      });
    })
  };

};

// Initialise a singleton Api Service 
export default new Api;
