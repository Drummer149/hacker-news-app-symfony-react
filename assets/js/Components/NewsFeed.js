import React from 'react';
import ReactDOM from 'react-dom';
import Api from '../Common/Api';
import InfiniteScroll from 'react-infinite-scroll-component';
import NewsItem from './NewsItem';

const style = {
    height: 30,
    border: "1px solid green",
    margin: 6,
    padding: 8
  };

const NUM_ITEMS_PAGE = 30;
const MAX_NUM_ITEMS = 500;


class NewsFeed extends React.Component {


    constructor() {
        super();

        this.state = {
            items: [],
            storyIds: [],
            endingIndex: 0,
            idsLoaded: false,
        };

        this.fetchMoreData = this.fetchMoreData.bind(this);
    }

    componentDidMount() {
        let startingIndex = 0;
        let endingIndex = startingIndex + NUM_ITEMS_PAGE;

        Api.getStoryIds(this.props.type)
            .then((storyIds) => {
                this.setState({storyIds: storyIds, idsLoaded: true});
                this.getData(startingIndex, endingIndex);
            })
        
    }

    getData(startingIndex, endingIndex){
        let actions = this.state.storyIds.slice(startingIndex, endingIndex).map(Api.getItem.bind(Api));
        // Resolve all the promises
        let results = Promise.all(actions);
        results.then(news =>
            {
                let newItems = this.state.items.concat(news);
                this.setState({items : newItems});
            });
    }

    fetchMoreData(){
        this.getData(this.state.items.length, this.state.items.length + NUM_ITEMS_PAGE);
    };

    render() {
        return (<InfiniteScroll
            className="newsFeed"
            dataLength={this.state.items.length}
            next={this.fetchMoreData}
            hasMore={this.state.items.length !== this.state.storyIds.length && this.state.idsLoaded}
            loader={<h4 style={{textAlign: 'center'}}>Loading...</h4>}
            endMessage={
                <p style={{textAlign: 'center'}}>
                  <b>You have reached the end of the internet</b>
                </p>
              }
          >
            {this.state.items.map((i, index) => (
              <NewsItem id={index} key={index} obj={i} />
            ))}
          </InfiniteScroll>)
    }
}

export default NewsFeed;