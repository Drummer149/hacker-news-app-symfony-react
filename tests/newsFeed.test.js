import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import React from 'react';
import { shallow } from 'enzyme';
import Api from '../assets/js/Common/Api';
import NewsFeed from '../assets/js/Components/NewsFeed';
import { mockStory, mockStoryIds } from './mockobjs';

describe('News Feed', () => {
  beforeAll(() => {
    const mockGetItem = () => {
      return new Promise(resolve => {
        resolve(mockStory);
      });
    };

    const mockGetStoryIds = () => {
      return new Promise(resolve => {
        resolve(mockStoryIds);
      });
    };

    jest.mock('../assets/js/Common/Api', () => {
        return {
          getStoryIds: mockGetStoryIds,
          getItem: mockGetItem,
        }
    });
  });

  const initialProps = {
    type: 'topstories'
  }

  it('should render', () => {
    const component = shallow(<NewsFeed {...initialProps} />);
    expect(component.exists()).toBe(true);
  });

  it('should render correctly', () => {
    const component = shallow(<NewsFeed {...initialProps} />);
    expect(component).toMatchSnapshot();
  });

  it('should populate the state with the StoryIds coming from the API', () => {
    jest.useFakeTimers();
    const component = shallow(<NewsFeed {...initialProps} />);

    setTimeout(() => {
      expect(component.state().storyIds).toBe(mockStoryIds);
    }, 1000);
  });
});