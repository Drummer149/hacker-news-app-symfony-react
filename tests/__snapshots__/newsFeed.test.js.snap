// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News Feed should render correctly 1`] = `ShallowWrapper {}`;
